strawberryfields/backends/tfbackend/circuit.py

```diff
@@ -540,139 +511,131 @@ def measure_fock(self, modes, select=None, **kwargs):
 
         # carry out the operation
-        with self.graph.as_default():
-            evaluate_results, session, feed_dict, close_session = ops._check_for_eval(kwargs)
-            num_reduced_state_modes = len(modes)
-            reduced_state = self._state
-            if self._state_is_pure:
-                mode_size = 1
-            else:
-                mode_size = 2
-            if self._batched:
-                batch_size = self._batch_size
-                batch_offset = 1
-            else:
-                batch_size = 1
-                batch_offset = 0
+        num_reduced_state_modes = len(modes)
+        reduced_state = self._state
+        if self._state_is_pure:
+            mode_size = 1
+        else:
+            mode_size = 2
+        if self._batched:
+            batch_size = self._batch_size
+            batch_offset = 1
+        else:
+            batch_size = 1
+            batch_offset = 0
 
-            if select is not None:
-                # just use the supplied measurement results
-                meas_result = select
+        if select is not None:
+            # just use the supplied measurement results
+            meas_result = select
+        else:
+            # compute and sample measurement result
+            if self._state_is_pure and len(modes) == self._num_modes:
+                # in this case, measure directly on the pure state
+                probs = tf.abs(self._state) ** 2
+                logprobs = tf.math.log(probs)
+                sample = tf.random.categorical(tf.reshape(logprobs, [batch_size, -1]), 1)
+                sample_tensor = tf.squeeze(sample)
             else:
-                # compute and sample measurement result
-                if self._state_is_pure and len(modes) == self._num_modes:
-                    # in this case, measure directly on the pure state
-                    probs = tf.abs(self._state) ** 2
-                    logprobs = tf.log(probs)
-                    sample = tf.multinomial(tf.reshape(logprobs, [batch_size, -1]), 1)
-                    sample_tensor = tf.squeeze(sample)
+                # otherwise, trace out unmeasured modes and sample using diagonal of reduced state
+                removed_ctr = 0
+                red_state_is_pure = self._state_is_pure
+                for m in range(self._num_modes):
+                    if m not in modes:
+                        new_mode_idx = m - removed_ctr
+                        reduced_state = ops.partial_trace(reduced_state, new_mode_idx, red_state_is_pure, self._batched)
+                        red_state_is_pure = False
+                        removed_ctr += 1
+                # go from bra_A,ket_A,bra_B,ket_B,... -> bra_A,bra_B,ket_A,ket_B,... since this is what diag_part expects
+                # workaround for getting multi-index diagonal since tensorflow doesn't support getting diag of more than one subsystem at once
```

Comment:
I'm not sure what you're asking

Reply:
The part where it says "since tensorflow doesn't support getting diag of more than one subsystem at once". I think this comment was written when TF 1.3 was being used, but maybe TF2.0 now supports this behaviour?